src/authentication/authentication.service.ts

```diff
@@ -44,4 +54,54 @@ export class AuthenticationService {
       token: this.generateAuthToken({ email: user.email, sub: user.id }),
     }
   }
+
+  async updatePasswordWithToken(
+    email: string,
+    token: string,
+    password: string,
+  ) {
+    let user
+    try {
+      this.jwtService.verify(token)
+      user = await this.usersService.findUserByResetToken(email, token)
+    } catch (e) {
+      if (
+        e instanceof SyntaxError || // token parse failed
+        e instanceof PrismaClientKnownRequestError // token doesn't match a user
+      ) {
+        throw new BadRequestException('Invalid token')
+      } else if (e instanceof TokenExpiredError) {
+        throw new BadRequestException('Token has expired')
```

Comment:
I think if the above is changed to `403`, then this conditional can go since either way, the token is just invalid due to exp or parsing..